### PR TITLE
Feat #833: Added Support for Applying Savings Account

### LINKED ISSUE #833: Support for applying Savings Accounts
**Details :**

Keeping consistency in the UI design throughout the app, helps in better UX. So, I’m thinking of
adding a Floating Action Button at the bottom of the fragment, similar to that in Loan Fragment.
Moreover, after adding Saving Accounts, we can also add a Confirmation Screen which will
show the details entered in the form, from there user can either Cancel or Apply from Saving
Accounts.

**Mockup :**

image

### PR DESCRIPTION
**Pull request by miPlodder**
Fixes #833 

For Adding Savings Account

pr_833_3

For Withdrawing and Updating Savings Account

pr_833_4

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

**Comment by miPlodder** [comment]
@therajanmaurya Please review this PR.

**Comment by miPlodder** [comment]
@therajanmaurya Withdrawal API are using Fake Responses and other are using the newly deployed API's and UI changed.

File: app/build.gradle [hunk 1]
```diff
@@ -44,6 +44,9 @@ android {
         test {
             java.srcDir commonTestDir
         }
+        main {
+            java.srcDir commonTestDir
```
Comment (therajanmaurya) [inline app/build.gradle]: Why this,  we already added the commonTestDir above.
See: https://github.com/openMF/mifos-mobile/pull/867/files#diff-39e7d8c00954e920b98e7636f0ac30b2R38

File: app/src/main/java/org/mifos/mobilebanking/models/accounts/savings/SavingsAccountWithdrawPayload.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package org.mifos.mobilebanking.models.accounts.savings;
+
+/*
+ * Created by saksham on 02/July/2018
+ */
+
+public class SavingsAccountWithdrawPayload {
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/models/accounts/savings/SavingsAccountWithdrawPayload.java]: Convert this payload into kotlin.

File: app/src/main/java/org/mifos/mobilebanking/models/templates/savings/ChargeOptions.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package org.mifos.mobilebanking.models.templates.savings
+
+import android.os.Parcelable
+import kotlinx.android.parcel.Parcelize
+import org.mifos.mobilebanking.models.accounts.savings.Currency
+
+@Parcelize
+data class ChargeOptions(
+
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/models/templates/savings/ChargeOptions.kt]: Remove this blank line.

File: app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountApplicationPresenter.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    @Override
+    public void detachView() {
+        super.detachView();
+        compositeDisposable.clear();
+    }
+
+    public void loadSavingsAccountApplicationTemplate(long clientId,
+                                                      final SavingsAccountState state) {
+        getMvpView().showProgress();
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountApplicationPresenter.java:54]: Check view `checkViewAttached();` is attached or not.

File: app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountApplicationPresenter.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import io.reactivex.disposables.CompositeDisposable;
+import io.reactivex.observers.DisposableObserver;
+import io.reactivex.schedulers.Schedulers;
+import okhttp3.ResponseBody;
+
+public class SavingsAccountApplicationPresenter
+        extends BasePresenter<SavingsAccountApplicationView> {
+
+    CompositeDisposable compositeDisposable;
+    DataManager dataManager;
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountApplicationPresenter.java]: Make this private

File: app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountApplicationPresenter.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.disposables.CompositeDisposable;
+import io.reactivex.observers.DisposableObserver;
+import io.reactivex.schedulers.Schedulers;
+import okhttp3.ResponseBody;
+
+public class SavingsAccountApplicationPresenter
+        extends BasePresenter<SavingsAccountApplicationView> {
+
+    CompositeDisposable compositeDisposable;
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountApplicationPresenter.java]: make this private

File: app/src/main/java/org/mifos/mobilebanking/models/templates/savings/SavingsAccountTemplate.kt [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+
+
+import android.os.Parcelable
+import kotlinx.android.parcel.Parcelize
+import java.util.ArrayList
+
+@Parcelize
+data class SavingsAccountTemplate(
+
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/models/templates/savings/SavingsAccountTemplate.kt]: Remove this blank line.

File: app/src/main/java/org/mifos/mobilebanking/models/templates/savings/ProductOptions.kt [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+/*
+ * Created by saksham on 01/July/2018
+ */
+
+@Parcelize
+data class ProductOptions(
+
+        var id: Int? = null,
+        var name: String? = null
+
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/models/templates/savings/ProductOptions.kt]: Remove this blank line.

File: app/src/main/java/org/mifos/mobilebanking/models/templates/savings/ProductOptions.kt [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import android.os.Parcelable
+import kotlinx.android.parcel.Parcelize
+
+/*
+ * Created by saksham on 01/July/2018
+ */
+
+@Parcelize
+data class ProductOptions(
+
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/models/templates/savings/ProductOptions.kt]: Remove this blank line.

File: app/src/main/java/org/mifos/mobilebanking/models/templates/savings/ChargePaymentMode.kt [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package org.mifos.mobilebanking.models.templates.savings
+
+import android.os.Parcelable
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class ChargePaymentMode (
+
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/models/templates/savings/ChargePaymentMode.kt]: Remove this blank line.

File: app/src/main/java/org/mifos/mobilebanking/models/templates/savings/ChargeOptions.kt (hunk 3 above)
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/models/templates/savings/ChargeOptions.kt]: Remove this blank line.

File: app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountApplicationPresenter.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+                    @Override
+                    public void onComplete() {
+
+                    }
+                }));
+    }
+
+    public void submitSavingsAccountApplication(SavingsAccountApplicationPayload payload) {
+        getMvpView().showProgress();
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountApplicationPresenter.java:86]: Check view `checkViewAttached();` is attached or not.

File: app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountApplicationPresenter.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+                    @Override
+                    public void onComplete() {
+                        getMvpView().showSavingsAccountApplicationSuccessfully();
+
+                    }
+                }));
+    }
+
+    public void updateSavingsAccount(String accountId, SavingsAccountUpdatePayload payload) {
+        getMvpView().showProgress();
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountApplicationPresenter.java:112]: Check view `checkViewAttached();` is attached or not.

File: app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountWithdrawPresenter.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.disposables.CompositeDisposable;
+import io.reactivex.observers.DisposableObserver;
+import io.reactivex.schedulers.Schedulers;
+import okhttp3.ResponseBody;
+
+public class SavingsAccountWithdrawPresenter extends BasePresenter<SavingsAccountWithdrawView> {
+
+    DataManager dataManager;
+    CompositeDisposable compositeDisposable;
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountWithdrawPresenter.java]: Make both `Composite....` `DataManager` private

File: app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountWithdrawPresenter.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Override
+    public void detachView() {
+        super.detachView();
+    }
+
+    public void submitWithdrawSavingsAccount(String accountId,
+                                             SavingsAccountWithdrawPayload payload) {
+        getMvpView().showProgress();
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountWithdrawPresenter.java:50]: Check view `checkViewAttached();` is attached or not.

File: app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountWithdrawPresenter.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Override
+    public void attachView(SavingsAccountWithdrawView mvpView) {
+        super.attachView(mvpView);
+    }
+
+    @Override
+    public void detachView() {
+        super.detachView();
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/presenters/SavingsAccountWithdrawPresenter.java:43]: Clear `CompositeDisposable` compositeDisposable.clear()

File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/SavingsAccountApplicationFragment.java [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+    @Inject
+    PreferencesHelper preferencesHelper;
+
+    View rootView;
+    SavingsAccountState state;
+    SavingsWithAssociations savingsWithAssociations;
+    ArrayAdapter<String> productIdAdapter;
+    SavingsAccountTemplate template;
+    List<ProductOptions> productOptions;
+    List<String> productIdList = new ArrayList<>();
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/ui/fragments/SavingsAccountApplicationFragment.java]: Make the above variables private

File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/SavingsAccountApplicationFragment.java [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+        payload.setSubmittedOnDate(DateHelper.getSpecificFormat(DateHelper.FORMAT_dd_MMMM_yyyy,
+                MFDatePicker.getDatePickedAsString()));
+        presenter.submitSavingsAccountApplication(payload);
+    }
+
+    public void updateSavingAccount() {
+        SavingsAccountUpdatePayload payload = new SavingsAccountUpdatePayload();
+        payload.setClientId(template.getClientId());
+        payload.setProductId(productOptions.get(spProductId.getSelectedItemPosition()).getId());
+
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/ui/fragments/SavingsAccountApplicationFragment.java]: Remove this blank line

File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/SavingsAccountWithdrawFragment.java [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+
+    @BindView(R.id.til_remark)
+    TextInputLayout tilRemark;
+
+    @Inject
+    SavingsAccountWithdrawPresenter presenter;
+
+    View rootView;
+    SavingsWithAssociations savingsWithAssociations;
+    SavingsAccountWithdrawPayload payload;
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobilebanking/ui/fragments/SavingsAccountWithdrawFragment.java]: Make above variables private.

**Review by therajanmaurya — CHANGES_REQUESTED** [review]

**Review by therajanmaurya — APPROVED** [review]

**Pull request merged**